<!-- move base -->
<launch>
  <master auto="start"/>
  <arg name="namespace"/>
  <arg name="configuration_basename" default="tdle.lua"/>
  <arg name="planner_name" default="teb"/>
  <param name="use_sim_time" value="true" />


<!-- using cartographer -->
  <node pkg="cartographer_ros" type="cartographer_node" name="cartographer_node" 
        args="-configuration_directory $(find tdle)/param
              -configuration_basename $(arg configuration_basename)"
        output="screen" launch-prefix="gnome-terminal --tab --" />  
  <node pkg="cartographer_ros" type="cartographer_occupancy_grid_node"
        name="cartographer_occupancy_grid_node" 
        args="-resolution 0.05" />

<!-- using Gmapping -->
   <!-- <node pkg="gmapping" type="slam_gmapping" name="slam_gmapping" output="screen" >
      <param name="map_frame" value="$(arg namespace)/map"/>
      <param name="odom_frame" value="$(arg namespace)/odom"/>
      <param name="base_frame" value="$(arg namespace)/base_link"/>
      <param name="map_update_interval" value="2.0"/>
      <param name="maxUrange" value="50.0"/>
      <param name="maxRange" value="50.0"/>
      <param name="sigma" value="0.05"/>
      <param name="kernelSize" value="1"/>
      <param name="lstep" value="0.05"/>
      <param name="astep" value="0.05"/>
      <param name="iterations" value="5"/>
      <param name="lsigma" value="0.075"/>
      <param name="ogain" value="3.0"/>
      <param name="lskip" value="10"/>
      <param name="srr" value="0.01"/>
      <param name="srt" value="0.02"/>
      <param name="str" value="0.01"/>
      <param name="stt" value="0.02"/>
      <param name="linearUpdate" value="0.01"/>
      <param name="angularUpdate" value="0.01"/>
      <param name="temporalUpdate" value="0.1"/>
      <param name="resampleThreshold" value="0.5"/>
      <param name="particles" value="30"/>
      <param name="xmin" value="-10.0"/>
      <param name="ymin" value="-10.0"/>
      <param name="xmax" value="10.0"/>
      <param name="ymax" value="10.0"/>
      <param name="delta" value="0.05"/>
      <param name="llsamplerange" value="0.01"/>
      <param name="llsamplestep" value="0.01"/>
      <param name="lasamplerange" value="0.005"/>
      <param name="lasamplestep" value="0.005"/>
      <param name="minimumScore" value="0.005"/>
    </node>  -->


<!-- using hector-slam -->
<!-- <node pkg="hector_mapping" type="hector_mapping" name="hector_mapping" output="screen">

    <param name="scan_topic" value="/$(arg namespace)/scan" />    

    <param name="map_frame" value="$(arg namespace)/map"/>
    <param name="odom_frame" value="$(arg namespace)/odom"/>
    <param name="base_frame" value="$(arg namespace)/base_link"/>
    <param name="output_timing" value="false"/>

    <param name="use_tf_scan_transformation" value="true"/>
    <param name="use_tf_pose_start_estimate" value="false"/>
    <param name="pub_map_odom_transform" value="true"/>
    <param name="advertise_map_service" value="true"/>

    <param name="map_resolution" value="0.050"/>
    <param name="map_size" value="300"/>
    <param name="map_start_x" value="0.5"/>
    <param name="map_start_y" value="0.5" />
    <param name="map_multi_res_levels" value="2" />

    <param name="update_factor_free" value="0.4"/>
    <param name="update_factor_occupied" value="0.9" />    
    <param name="map_update_distance_thresh" value="0.4"/>
    <param name="map_update_angle_thresh" value="0.06" />
  </node>
  <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="hector_trajectory_server" output="screen">
    <param name="target_frame_name" type="string" value="/map" />
    <param name="source_frame_name" type="string" value="/base_link" />
    <param name="trajectory_update_rate" type="double" value="4" />
    <param name="trajectory_publish_rate" type="double" value="0.25" />
  </node>
  <node pkg="hector_geotiff" type="geotiff_node" name="hector_geotiff_node" output="screen" launch-prefix="nice -n 15">
    <remap from="map" to="/dynamic_map" />
    <param name="map_file_path" type="string" value="$(find hector_geotiff)/maps" />
    <param name="map_file_base_name" type="string" value="hector_slam_map" />
    <param name="geotiff_save_period" type="double" value="10" />
    <param name="draw_background_checkerboard" type="bool" value="true" />
    <param name="draw_free_space_grid" type="bool" value="true" />
  </node> -->



  <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen" launch-prefix="gnome-terminal --tab --">
    <param name="footprint_padding" value="0.01" />
    <param name="controller_frequency" value="5.0" />
    <param name="controller_patience" value="3.0" />
    <param name="oscillation_timeout" value="30.0" />
    <param name="oscillation_distance" value="0.5" />
    <param name="planner_patience" value="1" />
    <param name="controller_patience" value="1" /> 
    <param name="recovery_behavior_enabled" value="true" />


    <param if="$(eval planner_name == 'teb')" name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
    <param if="$(eval planner_name == 'dwa')" name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>

    <rosparam file="$(find tdle)/param/$(arg planner_name)/costmap_common_params.yaml" command="load" ns="global_costmap" />
    <rosparam file="$(find tdle)/param/$(arg planner_name)/costmap_common_params.yaml" command="load" ns="local_costmap" />
    <rosparam file="$(find tdle)/param/$(arg planner_name)/local_costmap_params.yaml" command="load" />
    <rosparam file="$(find tdle)/param/$(arg planner_name)/global_costmap_params.yaml" command="load" />
    <rosparam file="$(find tdle)/param/$(arg planner_name)/local_planner_params.yaml" command="load" />
    
    <!-- frame -->
    <param name="global_costmap/global_frame" value="map"/>
    <param name="global_costmap/robot_base_frame" value="base_link"/>
    <param name="global_costmap/laser_scan_sensor/sensor_frame" value="base_scan"/>  
    <param name="local_costmap/global_frame" value="odom"/>
    <param name="local_costmap/robot_base_frame" value="base_link"/>
    <param name="local_costmap/laser_scan_sensor/sensor_frame" value="base_scan"/>
    
    <!-- topic -->
    <param name="global_costmap/laser_scan_sensor/topic" value="/$(arg namespace)/scan"/>  
    <param name="local_costmap/laser_scan_sensor/topic" value="/$(arg namespace)/scan"/>
    <param name="local_costmap/obstacle_layer/laser_scan_sensor/topic" value="/$(arg namespace)/scan"/>
  
  </node>

</launch>